src/ripple/protocol/impl/STAmount.cpp

```diff
@@ -1346,8 +1346,42 @@ canonicalizeRoundStrict(
     }
 }
 
-// Pass the canonicalizeRound function as a template parameter.
-template <void (*Canonicalize)(bool, std::uint64_t&, int&, bool)>
+namespace {
+
+// saveNumberRoundMode doesn't do quite enough for us.  What we want is a
+// Number::RoundModeGuard that sets the new mode and restores the old mode
+// when it leaves scope.  Since Number doesn't have that facility, we'll
+// build it here.
+class NumberRoundModeGuard
+{
+    saveNumberRoundMode saved_;
+
+public:
+    explicit NumberRoundModeGuard(Number::rounding_mode mode) noexcept
+        : saved_{Number::setround(mode)}
+    {
+    }
+};
+
+// We need a class that has an interface similar to NumberRoundModeGuard
+// but does nothing.
+class DontAffectNumberRoundMode
+{
+public:
```

Comment:
_Pedantic:_  To more exactly mimic the `NumberRoundModeGuard` API, consider deleting the special copy members:
```c++
    DontAffectNumberRoundMode(DontAffectNumberRoundMode const&) = delete;
    DontAffectNumberRoundMode& operator=(DontAffectNumberRoundMode const&) = delete;
```